apps/aecore/lib/aecore/chain/chainstate.ex

```diff
@@ -32,43 +32,60 @@ defmodule Aecore.Chain.Chainstate do
     }
   end
 
-  @spec calculate_and_validate_chain_state!(list(), Chainstate.t(), non_neg_integer()) ::
+  @spec calculate_and_validate_chain_state(list(), Chainstate.t(), non_neg_integer()) ::
           Chainstate.t()
-  def calculate_and_validate_chain_state!(txs, chainstate, block_height) do
+  def calculate_and_validate_chain_state(txs, chainstate, block_height) do
     txs
     |> Enum.reduce(chainstate, fn tx, chainstate ->
-      apply_transaction_on_state!(tx, chainstate, block_height)
+      case apply_transaction_on_state(tx, chainstate, block_height) do
+        {:ok, updated_chainstate} ->
+          updated_chainstate
+
+        {:error, _reason} ->
+          chainstate
```

Comment:
(important) If any transaction fails this whole function has to fail.

Reply:
(Failing transactions shouldn't be skipped, it should cause block validation failure)

Reply:
I think there is a fundamental problem with the use of this function. First I don't see why this function would crash inside the [`miner`](https://github.com/aeternity/elixir-node/blob/03880f4a5ab5fa25ad0033809827618e4be1d500/apps/aecore/lib/aecore/miner/worker.ex#L311) because we pass already filtered transactions to it. So why would we need to check them again?

If the crashing is ment for the [`BlockValiedation`]() usage then maybe this should be a seperate func.

I'll give you a roadmap of what we are doing atm and you tell me what you think about it. I find it strange to say the least:
1st -> We call [`Miner.mine_sync_block_to_chain()`](https://github.com/aeternity/elixir-node/blob/03880f4a5ab5fa25ad0033809827618e4be1d500/apps/aecore/lib/aecore/miner/worker.ex#L77)
2nd -> We call [`Miner.candidate()`](https://github.com/aeternity/elixir-node/blob/03880f4a5ab5fa25ad0033809827618e4be1d500/apps/aecore/lib/aecore/miner/worker.ex#L225)
3rd -> We [`filter the invalid txs`](https://github.com/aeternity/elixir-node/blob/03880f4a5ab5fa25ad0033809827618e4be1d500/apps/aecore/lib/aecore/miner/worker.ex#L244)
4th -> We call [`Miner.create_block`](https://github.com/aeternity/elixir-node/blob/03880f4a5ab5fa25ad0033809827618e4be1d500/apps/aecore/lib/aecore/miner/worker.ex#L307) passing those valid txs which we validated by fee and added the coinbase tx as well
5th -> We call [`Chain.calculate_and_validate_chain_state()`](https://github.com/aeternity/elixir-node/blob/03880f4a5ab5fa25ad0033809827618e4be1d500/apps/aecore/lib/aecore/miner/worker.ex#L310) which I don't see a reason to fail having in mind that the txs are already filtered and shouldn't crash the function
6th -> We get the returned block from the `Miner.candidate()` func and we pass it to [`Chain.add_block`](https://github.com/aeternity/elixir-node/blob/03880f4a5ab5fa25ad0033809827618e4be1d500/apps/aecore/lib/aecore/miner/worker.ex#L81)
7th -> Inside the Chain.add_block() func we call [`BlockValidation.calculate_and_validate_block()`](https://github.com/aeternity/elixir-node/blob/09c7fe3331727e278030d923328345bb8ed03a06/apps/aecore/lib/aecore/chain/block_validation.ex#L26) which inside itself calls `Chain.calculate_and_validate_chain_state()` again, where we pass the same filtered transactions we got from `Miner.candidate()` func. And then again why do we need to call `Chain.calculate_and_validate_chain_state()` when we just did so inside the [`Miner.create_block()`](https://github.com/aeternity/elixir-node/blob/03880f4a5ab5fa25ad0033809827618e4be1d500/apps/aecore/lib/aecore/miner/worker.ex#L307) func

So I don't see any reason why `Chain.calculate_and_validate_chain_state()` would creash. If it does then the problem is in the filtering of the invalid txs, because when we filter the invalid txs we call the same function we call inside the `Chain.calculate_and_validate_chain_state()` which is `Chain.apply_transaction_on_state()`.

Correct me if I'm missing something here.

Reply:
You are missing that the block may come from the network and Chain.add_block is supposed to validate the block before adding it. (The block isn't fully validated before reaching this function)

`Chain.calculate_and_validate_chain_state()` implements our full validation of transactions logic. Before if any transaction would be invalid for any reason this function would throw. Now I understand we want to use return values instead of throwing due to performance reasons. So we should return something like `:error` or `{:error, reason}` in case of failure of any transaction.

As for the miner: the miner validates new blocks twice or maybe three times. This could be optimized, but executing different logic for blocks from miner and those from network may create harder to debug problems (if miner creates invalid blocks). The overhead of validating blocks twice should be negligible in practice compared to cost of running miner. Even in our tests.

As I said briefly in the oryginal comment: block with any invalid transaction is invalid. So failure (of one of the checks) when applying one of the transcriptions to chainstate should cause whole block validation to fail.

Reply:
Well okay, then why don't we change the `Chain.calculate_and_validate_chain_state()` to not check the txs having in mind that we already have such a function. Duplicating code is not a good approach either. We could then chain the functions to go after one another in the `BlockValidation.validate_and_calculate_block()` and don't check twice the txs in the Miner. Another idea is to add an atom to the `BlockValidation` func to check weather we are comming from the miner of from the network and not call the same funcs again. But I think this should be a seperate issue having in mind that it will change some backbone code. It should be more visible than doing it in this branch.

Reply:
What checking function do you have in mind? There are 3 functions executions for each transaction (and I don't think we should add more):
- `is_valid?` - checks transactions internals
- `preprocess_check` - checks if transaction can execute on top of given chainstate
- `process_chainstate` - executes transaction of top of given chainstate

Generally it's not possible to validate transactions without executing them due to dependencies.

It's not possible to validate a list of transactions without executing them due to potential dependencies. If you have a list of transactions [T1,T2, T3, T4, ...] and a chainstate C. To validate T3 you have to apply T1 and T2 to C and execute preporocess_check with resulting chainstate on T3. Due to that validating a list of transactions has a side effect of creating chainstate with all of the transactions executed.

As for the Miner again: I think even currently there exists a function in `Aecore.Chain` where you can supply already validated block to be included.

Reply:
If we call `Chain.calculate_and_validate_chain_state()` twice in some flow then we can fix that. As for the return value: if we would remove returning errors from `Chain.calculate_and_validate_chain_state()`, then we would have to calculate transactions twice for validation of blocks from network.